File: src/chipschallenge/blockreactions/RecessedWallTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.BlockContainerFullException;

public class RecessedWallTo extends NoSlipReaction {

    private RecessedWallTo() {
    }
    private static RecessedWallTo mInstance = null;

    public static synchronized RecessedWallTo getInstance() {
        if (mInstance == null) {
            mInstance = new RecessedWallTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) throws BlockContainerFullException {
        standing.replace(createBlock(Type.WALL));
    }

    public boolean canMove(Block moving, Block standing) {
        return moving.isChip();
    }
}

File: ChipsChallenge.Shared/Blockreactions/RecessedWallTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Type = Block.Type;

    public class RecessedWallTo : NoSlipReaction
    {
        private static RecessedWallTo instance;

        private RecessedWallTo()
        {
        }

        public static RecessedWallTo Instance
        {
            get
            {
                lock (typeof(RecessedWallTo))
                {
                    return instance ?? (instance = new RecessedWallTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            standing.Replace(CreateBlock(Type.WALL));
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip;
        }
    }
}